pkg/codec/vpx/vpx.go

```diff
@@ -0,0 +1,182 @@
+	mio "github.com/pion/mediadevices/pkg/io"
+	"github.com/pion/mediadevices/pkg/io/video"
+	"github.com/pion/mediadevices/pkg/prop"
+
+	"github.com/pion/webrtc/v2"
+)
+
+type encoder struct {
+	codec            *C.vpx_codec_ctx_t
+	raw              *C.vpx_image_t
+	r                video.Reader
+	frameIndex       int
+	keyframeInterval int
+	buff             []byte
+}
+
+func init() {
+	codec.Register(webrtc.VP8, codec.VideoEncoderBuilder(NewVP8Encoder))
+	codec.Register(webrtc.VP9, codec.VideoEncoderBuilder(NewVP9Encoder))
+}
+
+// NewVP8Encoder creates new VP8 encoder
+func NewVP8Encoder(r video.Reader, p prop.Video) (io.ReadCloser, error) {
+	return newEncoder(r, p, C.ifaceVP8())
+}
+
+// NewVP9Encoder creates new VP9 encoder
+func NewVP9Encoder(r video.Reader, p prop.Video) (io.ReadCloser, error) {
+	return newEncoder(r, p, C.ifaceVP9())
+}
+
+func newEncoder(r video.Reader, p prop.Video, codecIface *C.vpx_codec_iface_t) (io.ReadCloser, error) {
+	cfg := &C.vpx_codec_enc_cfg_t{}
+	if ec := C.vpx_codec_enc_config_default(codecIface, cfg, 0); ec != 0 {
+		return nil, fmt.Errorf("vpx_codec_enc_config_default failed (%d)", ec)
+	}
+	cfg.g_w = C.uint(p.Width)
+	cfg.g_h = C.uint(p.Height)
+	cfg.g_timebase.num = 1
+	cfg.g_timebase.den = 30 // TODO: p.FrameRate should be set
+	cfg.rc_target_bitrate = C.uint(p.BitRate) / 1000
+
+	raw := &C.vpx_image_t{}
+	if C.vpx_img_alloc(raw, C.VPX_IMG_FMT_I420, cfg.g_w, cfg.g_h, 1) == nil {
+		return nil, errors.New("vpx_img_alloc failed")
+	}
+	rawNoBuffer := C.newImage()
+	*rawNoBuffer = *raw // Copy only parameters
+	C.vpx_img_free(raw) // Pointers will be overwritten by the raw buffer
+
+	codec := C.newCtx()
+	if ec := C.vpx_codec_enc_init_ver(
+		codec, codecIface, cfg, 0, C.VPX_ENCODER_ABI_VERSION,
+	); ec != 0 {
+		return nil, fmt.Errorf("vpx_codec_enc_init failed (%d)", ec)
+	}
+	return &encoder{
+		r:                r,
+		codec:            codec,
+		raw:              rawNoBuffer,
+		keyframeInterval: 30, // TODO: Set via prop.Video
+	}, nil
+}
+
+func (e *encoder) Read(p []byte) (int, error) {
+	if e.buff != nil {
+		n, err := mio.Copy(p, e.buff)
+		if err == nil {
+			e.buff = nil
+		}
+		return n, err
+	}
+
+	img, err := e.r.Read()
+	yuvImg := img.(*image.YCbCr)
+
+	e.raw.planes[0] = (*C.uchar)(&yuvImg.Y[0])
+	e.raw.planes[1] = (*C.uchar)(&yuvImg.Cb[1])
+	e.raw.planes[2] = (*C.uchar)(&yuvImg.Cr[2])
+	e.raw.stride[0] = C.int(yuvImg.YStride)
+	e.raw.stride[1] = C.int(yuvImg.CStride)
+	e.raw.stride[2] = C.int(yuvImg.CStride)
+	/*
+		// TODO: yuvImg.SubsampleRatio should be set
+		switch yuvImg.SubsampleRatio {
+		case image.YCbCrSubsampleRatio444:
+			e.raw.fmt = C.VPX_IMG_FMT_I444
+		case image.YCbCrSubsampleRatio422:
+			e.raw.fmt = C.VPX_IMG_FMT_I422
+		case image.YCbCrSubsampleRatio420:
+			e.raw.fmt = C.VPX_IMG_FMT_I420
+		case image.YCbCrSubsampleRatio440:
+			e.raw.fmt = C.VPX_IMG_FMT_I440
+		default:
+			return 0, fmt.Errorf("unsupported pixel format: %s", yuvImg.SubsampleRatio)
+		}
```

Comment:
Is this probably the reason why the color is off? Just curious.

Reply:
VP8 looks only supporting I420 and YV12. (VP9 supports I422, I444 and I440 on profile1, but pion/rtp doesn't have payloader yet.)
On my environment, image from camera is I422.

I'll add convert to make it work.

Reply:
Ah, I see.